text/3748-cargo_check_environment_variable.md

```diff
@@ -0,0 +1,107 @@
+- Feature Name: `cargo_check_environment_variable`
+- Start Date: 2024-12-20
+- RFC PR: [rust-lang/rfcs#3748](https://github.com/rust-lang/rfcs/pull/3748)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a new environment variable `CARGO_CHECK` that is set to `true` when running `cargo check` or similar type-checking operations so build scripts can skip expensive compilation steps that are unnecessary for Rust type checking, such as compiling external C++ code in cxx based projects.
+
+# Motivation
+[motivation]: #motivation
+
+Rust development heavily relies on IDE tooling like rust-analyzer, which frequently invokes `cargo check` to provide real-time type information and diagnostics. Many projects use build scripts (`build.rs`) to generate Rust code and compile external dependencies. For example:
+
+- cxx-rs generates Rust bindings for C++ code and compiles C++ source files
+- cxx-qt generates Rust bindings for Qt code and runs the Qt Meta-Object Compiler (MOC)
+- Projects using Protocol Buffers generate Rust code from .proto files
+- bindgen generates Rust bindings from C/C++ headers
+
+Currently, every time rust-analyzer runs `cargo check`, all build scripts must execute their full build process, including steps like compiling C++ code that are only needed for linking but not for type checking. This significantly impacts IDE responsiveness, especially in projects with complex build scripts.
+
+This is particularly important for projects using cxx-qt and similar frameworks where the build scripts perform extensive code generation and compilation.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When writing a build script (`build.rs`), you can now check the `CARGO_CHECK` environment variable to determine if the build is being performed for type checking purposes:
+
+```rust
+fn main() {
+    generate_rust_bindings();
+
+    // Only compile external code when not type checking
+    if std::env::var("CARGO_CHECK").unwrap() != "true" {
+        compile_cpp_code();
+    }
+}
+```
+
+This allows build scripts to optimize their behavior based on the build context. When rust-analyzer or a developer runs `cargo check`, the build script can skip time-consuming steps that aren't necessary for type checking.
+
+This feature primarily benefits library authors who maintain build scripts, especially those working with external code generation and compilation. Regular Rust developers using these libraries will automatically benefit from improved IDE performance without needing to modify their code.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Cargo will set the `CARGO_CHECK` environment variable to `true` when running `cargo check`
```

Comment:
I think it is unwise to name and describe this as for `cargo check` in particular, for two reasons:

* A future-compatibility hazard, especially in context of extending this to other modes as noted in the future possibilities section — are build scripts going to have to check for `CARGO_CHECK || CARGO_DOC || CARGO_CLIPPY` to be efficient? I think we’d see future arguments over whether commands are “check-like enough to count” rather than introducing more, which would spread confusion and create a tradeoff we can just avoid from the start.
* Build script authors might decide to intentionally do non-additive things based on the command in use (like generating different bindings code that’s more development-friendly in some way), leading to incorrect caching.

I think it would be better to name and describe as something like:

> Cargo will set the `CARGO_WILL_NOT_LINK` environment variable whenever Cargo’s build plan does not perform any linking (and therefore is known not to depend on any foreign object files the build script might produce). Whenever `CARGO_WILL_NOT_LINK` is present, you may skip producing object files. All generated Rust code or other inputs to the Rust compilation (such as files for `include_str!` or `cargo::rustc-cfg` directives) must be unaffected.
> 
> The value of `CARGO_WILL_NOT_LINK` is reserved for making finer distinctions in the future; currently, it is always set to `"1"`, but build scripts should check only for its presence.

This does change the behavior because the definition includes all of `cargo check`, `cargo fix`, `cargo doc`, and `cargo clippy`, but I think that’s the right thing to do. It’s also very specific about what the build script may do and should not do, to ensure that the results of the build script can still validly be cached.

(I specified the value to be `"1"` and not `"true"` because [that's what Cargo uses for feature flags](https://github.com/rust-lang/cargo/blob/652623b779c88fe44afede28bf7f1c9c07812511/src/cargo/core/compiler/custom_build.rs#L335), the other existing example of “boolean” environment variables.)

Reply:
These are great points!

I think your alternative is a better approach, especially not to step on the toes of the parallel cargo modes discussion. I'll wait for more input and then probably update the RFC to the alternative method after the holidays

Reply:
There are a handful of environment variables that don't have the `CARGO_` prefix, e.g. `OUT_DIR` `TARGET` `NUM_JOBS` `OPT_LEVEL`. My read is that these unprefixed variables are loosely meant to be portable to other build systems that might want to reuse build scripts. Since the proposed feature is certainly something that other build systems could make use of, maybe it should also be unprefixed?

A name like `CHECK_ONLY` would be reasonably intuitive. `WILL_NOT_LINK` or `SKIP_CODEGEN` are a bit more technical but maybe more accurate. `NO_CODEGEN` is consistent with the unstable `rustc -Zno-codegen`. I think Cargo currently gets the check-only functionality by passing `--emit=dep-info,metadata`.

List of the existing environment variables for reference https://doc.rust-lang.org/cargo/reference/environment-variables.html